File: src/leetcode/easy/sameTree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

class TreeNode {
	val: number;
	left: TreeNode | null;
	right: TreeNode | null;
	constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
		this.val = val === undefined ? 0 : val;
		this.left = left === undefined ? null : left;
		this.right = right === undefined ? null : right;
	}
}

const isSameTree = (p: TreeNode | null, q: TreeNode | null): boolean => {
	const dfs = (currNode: TreeNode | null): string => {
		if (currNode === null) {
			return 'null';
		} else {
			return currNode.val.toString() + dfs(currNode.left) + dfs(currNode.right);
		}
	};

	return dfs(p) === dfs(q);
};

File: src/leetcode/easy/sameTree.py
from typing import Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        queue = [[p, q]]

        while len(queue) > 0:
            curr_node = queue.pop(0)

            [curr_p, curr_q] = curr_node

            if curr_p or curr_q:
                if curr_p is None or curr_q is None or curr_p.val != curr_q.val:
                    return False

                queue.append([curr_p.left, curr_q.left])
                queue.append([curr_p.right, curr_q.right])

        return True
